app/code/Bulbulatory/Recommendations/Controller/Adminhtml/Table/DeleteMass.php

```diff
@@ -0,0 +1,86 @@
+<?
+
+namespace Bulbulatory\Recommendations\Controller\Adminhtml\Table;
+
+use Bulbulatory\Recommendations\Model\ResourceModel\Recommendation\CollectionFactory;
+use Magento\Backend\App\Action;
+use Magento\Backend\App\Action\Context;
+use Magento\Ui\Component\MassAction\Filter;
+use Bulbulatory\Recommendations\Model\RecommendationRepository;
+
+/**
+ * Class DeleteMass
+ * @package Bulbulatory\Recommendations\Controller\Adminhtml\Table
+ */
+class DeleteMass extends Action
+{
+    /**
+     * @var Filter
+     */
+    protected $filter;
+
+    /**
+     * @var CollectionFactory
+     */
+    protected $collectionFactory;
+
+    /**
+     * @var RecommendationRepository
+     */
+    protected $recommendationRepository;
+
+    /**
+     * DeleteMass constructor.
+     * @param Filter $filter
+     * @param CollectionFactory $collectionFactory
+     * @param Context $context
+     * @param RecommendationRepository $recommendationRepository
+     */
+    public function __construct(
+        Filter $filter,
+        CollectionFactory $collectionFactory,
+        Context $context,
+        RecommendationRepository $recommendationRepository
+    )
+    {
+        $this->filter = $filter;
+        $this->collectionFactory = $collectionFactory;
+        $this->recommendationRepository = $recommendationRepository;
+
+        parent::__construct($context);
+    }
+
+    /**
+     * @return mixed
+     */
+    protected function _isAllowed()
+    {
+        return $this->_authorization->isAllowed('Bulbulatory_Recommendations::bulbulatory_recommendations_table_delete');
+    }
+
+    /**
+     * @return mixed
+     */
+    public function execute()
+    {
+        $resultRedirect = $this->resultRedirectFactory->create();
+
+        try {
+            $recommendationCollection = $this->filter->getCollection($this->collectionFactory->create());
+
+            foreach ($recommendationCollection as $item)
+            {
+                $id = $item->getId();
+                $recommendation = $this->recommendationRepository->getById($id);
+                $recommendation->delete();
+                $this->messageManager->addSuccessMessage(__('Recommendation with ID: ' . $id . ' deleted!'));
+            }
+        }
+        catch (\Throwable $e)
+        {
+            $this->messageManager->addErrorMessage($e->getMessage());

```

Comment:
You have a try/catch block covering the whole foreach which means that one exceptions occurs and the execution stops. When you cover only the lines with fetching the recommendation and deleting it, you can log the exception and the rest of the loop will execute anyway. This way when one recommendation is problematic for some reason, all the rest can be strill removed.

Reply:
@TM18

> You have a try/catch block covering the whole foreach which means that one exceptions occurs and the execution stops. When you cover only the lines with fetching the recommendation and deleting it, you can log the exception and the rest of the loop will execute anyway. This way when one recommendation is problematic for some reason, all the rest can be strill removed.

You're right! Fixed!